### PR TITLE
Implement FreezeService preHandle

### PR DESCRIPTION
**Pull request by kimbor**
Implement FreezeService preHandle for Admin module

Issue https://github.com/hashgraph/hedera-services/issues/4464

### CODE REVIEW

**Review by netopyr — APPROVED** [review]
LGTM from a hedera-app perspective.

File: hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/FreezeServiceImpl.java
Comment (iwsimon) [inline hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/FreezeServiceImpl.java:1]: Why do we use SPECIAL_FILES_KEY for the stateKey on this freeze service?

File: hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/ReadableSpecialFileStore.java
Comment (iwsimon) [inline hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/ReadableSpecialFileStore.java:1]: ReadableKVState<Long, byte[]>, should the byte[] be a object type?

File: hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java
Comment (iwsimon) [inline hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java:1]: We should just throw PreCheckExceptions in preHandle() and let PreHandleWorkflowImpl.preHandle to catch those exceptions and handle them there.

Comment (kimbor) [reply hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java:1]: Yes, we had a discussion last week that came to that conclusion. None of the services do that now but we decided that they should all be updated to do that. I created issue 5880 for this. For the time being, until PreHandleWorkflowImpl is updated, this PR goes along with what all the other services are doing.

File: hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/FreezeServiceImpl.java
Comment (kimbor) [reply hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/FreezeServiceImpl.java:1]: I moved that over from NetworkService and didn't rename it. I think it's called that for historical reasons that may not be valid any more. Perhaps UPGRADE_FILE_KEY would be a better name.

Comment (kimbor) [reply hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/FreezeServiceImpl.java:1]: I renamed it to UPGRADE_FILES_KEY

**Review by netopyr — APPROVED** [review]
LGTM from hedera-app perspective

**Review by iwsimon — APPROVED** [review]
LGTM

**Review by povolev15 — APPROVED** [review]
LGTM

**Review by qnswirlds — APPROVED** [review]

File: hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java [hunk 1]
```diff
@@ -31,26 +43,125 @@
...
+    // it is necessary to check getStartHour, getStartMin, getEndHour, getEndMin, all of which are deprecated
+    // because if any are present then we set a status of INVALID_FREEZE_TRANSACTION_BODY
+    public void preHandle(
+            @NonNull final PreHandleContext context, @NonNull final ReadableSpecialFileStore specialFileStore) {
         requireNonNull(context);
-        throw new UnsupportedOperationException("Not implemented");
+
+        FreezeTransactionBody freezeTxn = context.getTxn().freeze();
+
+        try {
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java:71]: I would suggest all validation go into a method called `validate` or `preCheck` or similar. And we call that method as first step in preHandle

File: hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java [hunk 2]
```diff
@@ -31,26 +43,125 @@
...
+                case FREEZE_ABORT -> {
+                    // do nothing
+                }
+            }
+
+            // no need to add any keys to the context because this transaction does not require any signatures
+            // it must be submitted by an account with superuser privileges, that is checked during ingest
+
+            // all checks have passed
+            context.status(OK);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java:118]: I believe the default status is OK if nothing is set. So, I think this is not needed

File: hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java [hunk 3]
```diff
@@ -31,26 +43,125 @@
...
+        } catch (PreCheckException e) {
+            // instead of catching this exception, would like to allow it to propagate up
+            // this will be implemented in issue #5880
+            context.status(e.responseCode());
+        }
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java:123]: We should not be catching any exceptions and setting status on context. It should be thrown as `PrecheckException`

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/ReadableStoreFactory.java [hunk 4]
```diff
@@ -72,6 +74,17 @@ public ReadableScheduleStore createScheduleStore() {
         return new ReadableScheduleStore(scheduleStates);
     }
 
+    /**
+     * Get a {@link ReadableScheduleStore}
+     *
+     * @return a new {@link ReadableScheduleStore}
+     */
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/ReadableStoreFactory.java:81]: Typo ReadableSpecialFileStore

**Review by Neeharika-Sompalli — COMMENTED** [review]
Review post merge.

**Pull request merged**